opal/mca/btl/vader/btl_vader_component.c

```diff
@@ -350,6 +350,48 @@ static int mca_btl_vader_component_close(void)
     return OPAL_SUCCESS;
 }
 
+/*
+ * mca_btl_vader_parse_proc_ns_user() tries to parse the user namespace ID
+ * of the current process.
+ * Returns the ID of the user namespace. In the case of an error '0' is returned.
+ */
+uint64_t mca_btl_vader_parse_proc_ns_user(void)
+{
+    char *link = malloc(PATH_MAX);
+    pid_t pid = getpid();
+    uint64_t user_ns_id;
+    char *tmp;
+    int i;
+
+    i = readlink("/proc/self/ns/user", link, PATH_MAX);
+    if (-1 == i) {
+        return 0;
+    }
+
+    /*
+     * Result in link should look like 'user:[<inode-number>]', so at least
+     * 8 characters long: 'user:[?]'.
+     */
+    if (8 > i) {
+        free(link);
+        opal_output(0, "Error reading user namespace ID of process %d\n", pid);
+        return 0;
+    }
+
+    /* remove trailing ']' */
+    link[i - 1] = '\0';
+    tmp = strchr(link, '[');
+    if (NULL == tmp) {
+        free(link);
+        opal_output(0, "Error reading user namespace ID of process %d\n", pid);
```

Comment:
I'm sorry -- one more nit pick...

If this error occurs, it is likely to happen N times on a single node (i.e., for all MPI procs launched on the same node).  If the cluster is homogeneous, it'll happen for every single MPI process in the job.

As such, this `opal_output()` (and the one above it) should be `opal_show_help()`, because `opal_show_help()` de-duplicates (i.e., it'll only show it in the output of `mpirun` once, even if multiple processes emit the same help message).

That being said, do we know if the format of the `/proc/self/ns/user` link has always been the same (`user:[<inode-number>]`)?  I.e., is it possible we're running on an older Linux kernel with a different sym link format, and the format is *not* actually an error?

This is a logical question to ask since you handle the case of having no sym link by simply returning 0 without emitting any kind of error message.

Reply:
> As such, this opal_output() (and the one above it) should be opal_show_help(), because opal_show_help() de-duplicates (i.e., it'll only show it in the output of mpirun once, even if multiple processes emit the same help message).

Will change it.

> That being said, do we know if the format of the /proc/self/ns/user link has always been the same (user:[<inode-number>])? I.e., is it possible we're running on an older Linux kernel with a different sym link format, and the format is not actually an error?

With the changes from @ggouaillardet included to use `stat()` this is probably irrelevant.